rclcpp_lifecycle/src/transition.cpp

```diff
@@ -36,54 +50,109 @@ Transition::Transition(uint8_t id, const std::string & label)
 
 Transition::Transition(
   uint8_t id, const std::string & label,
-  State && start, State && goal)
-: owns_rcl_transition_handle_(true)
+  State && start, State && goal,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_transition_handle_(true),
+  transition_handle_(nullptr)
 {
-  auto transition_handle = new rcl_lifecycle_transition_t;
+  auto transition_handle = reinterpret_cast<rcl_lifecycle_transition_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_transition_t), allocator_.state));
+  if (!transition_handle) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_transition_t");
+  }
   transition_handle->id = id;
-  transition_handle->label = label.c_str();
+  transition_handle->label = rcutils_strndup(
+    label.c_str(), label.size(), allocator_);
 
-  auto start_state = new rcl_lifecycle_state_t;
+  auto start_state = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
+  if (!start_state) {
+    throw std::runtime_error("failed to allocate memory for rcl_lifecycle_state_t");
```

Comment:
This would leak `transition_handle`.

Same below.

Reply:
I added the `reset()` function in this PR and call it whenever one of the allocation fails.

Reply:
I think the allocated handles are not fully initialized with zeros therefore the `reset()` function will not be able to detect if a specific field needs to be deallocated since it might contain non-null values without containing actual data.

Reply:
> I think the allocated handles are not fully initialized with zeros therefore the reset() function will not be able to detect if a specific field needs to be deallocated since it might contain non-null values without containing actual data.

I am not sure where I noticed this. Looking again at the current state of the patch all members of the `*_state_` variables are initialized. So please ignore this comment.